### PR TITLE
Feature/#46 observe mounts

### LINKED ISSUE #46: Handle already-unmounted volumes gracefully
Calling `umount` on an already unmounted volume fails and the error cascades up and breaks vault locking, even though an already unmounted volume is an acceptable error. See https://github.com/cryptomator/cryptomator/issues/1192 for full details in context–I'm filing this issue with just the context that is relevant to fuse-nio-adapter. 

I had some energy so I hunted down this bug. I'm not set up for Java dev right now so I'm limited in my ability to submit a PR, and debugging java codebases with github's web interface is a huge pain, but the issue seems to ultimately come from how you are incorrectly checking whether the volume is already mounted or not.

AbstractMount does try to check first whether or not the Fuse volume is mounted, calling the isMounted interface of FuseNioAdapter:
https://github.com/cryptomator/fuse-nio-adapter/blob/4bb19ad6dd99598c6502944eedf4988445ba48a9/src/main/java/org/cryptomator/frontend/fuse/mount/AbstractMount.java#L47

Which just gets an AtomicBoolean that gets managed internally: https://github.com/cryptomator/fuse-nio-adapter/blob/4bb19ad6dd99598c6502944eedf4988445ba48a9/src/main/java/org/cryptomator/frontend/fuse/ReadOnlyAdapter.java#L258
https://github.com/cryptomator/fuse-nio-adapter/blob/4bb19ad6dd99598c6502944eedf4988445ba48a9/src/main/java/org/cryptomator/frontend/fuse/ReadOnlyAdapter.java#L268

That means that when a volume is unmounted outside of these java methods (which can easily happen for many many different and entirely valid reasons), the mounted flag doesn't have its state updated to match, and isMounted() doesn't check the actual reality of whether the volume is mounted or not. When Cryptomator is trying to unmount a volume, it does correctly try to check to avoid calling `umount` on an already unmounted volume, but that check as implemented can return a value inconsistent with the actual mounted state of the volume. In this case, the check will return true, and cryptomator will proceed to try to `umount` a volume that isn't mounted, resulting in an error.

It seems like one way to frame and approach this is as not handling the `umount` error specifically enough (because "not currently mounted" actually implies success for this particular context, and the rest of vault locking should be able to proceed as normal). The unmount syscall is meant to return EINVAL for this failure, unfortunately on OS X neither `umount` nor `diskutil unmount` seem to return useful error codes, so you'd be left parsing the error message:

```
% umount -f /Volume/fake; echo $?
umount: /Volume/fake: not currently mounted
1
% umount /Volume/fake; echo $?
umount: /Volume/fake: not currently mounted
1
```

Another way to frame it is that you need to check the actual mounted status before trying to call `umount`. On OS X	I don't know of a better way than parsing the output of `mount`. You could try something as simple as `mount | /usr/bin/awk '$3 == "/Volumes/path/to/volume" {print $3}'` (or do the parsing in Java, w/e). Either way is going to be platform specific so it seems like this checking behavior probably belongs in the Mount interface.

### PR DESCRIPTION
**Pull request by infeo**
Closes #46.

This PR adds the capability to register a task during mount(...), which is executed when the native fuse_main() method exits (regularly or irregularly).

## Implementation Details
A new abstract class `AbstractMounter` is added implementing the `Mounter.mount(...)` method and from which all MountImpls extend.

The mount method is extended with the parameter `Consumer<Throwable> onFuseExit`. The consumer is called after the blocking fuse call exits. If an exception is raised in the glue code around fuse_main(), it is set as the parameter to the consumer. (otherwise null)

In the mount method, a new Thread with the blocking `fuse_main()` call is spawned. The given fuse filesystem is wrapped with `MountSuccess`-Decorator, which releases a barrier in the FUSE `init()` method. The regular thread waits on this barrier 3s and on a timeout/interrupt/exception throws an exception.

Finally, the API exception `CommandFailedException` is renamed to `FuseMountException` and used as the general exception the API throws.

### CODE REVIEW

File: src/main/java/org/cryptomator/frontend/fuse/mount/AbstractMounter.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        mountThread.setDaemon(true);
+        mountThread.start();
+
+        // wait for mounted() is called, unlocking the barrier
+        try {
+            if (!mountSuccessSignal.await(MOUNT_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS)) {
+                if (exception.get() != null) {
+                    throw new FuseMountException(exception.get());
+                }
+                throw new FuseMountException("Mounting operation timed out");
```
Comment (overheadhunter) [inline src/main/java/org/cryptomator/frontend/fuse/mount/AbstractMounter.java:46]: wrap in else, so one can immediately see that both branches are mutually exclusive

File: src/main/java/org/cryptomator/frontend/fuse/mount/MountSuccessFuseDecorator.java [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+import jnr.ffi.Pointer;
+import org.cryptomator.frontend.fuse.FuseNioAdapter;
+import ru.serce.jnrfuse.FuseFS;
+import ru.serce.jnrfuse.FuseFillDir;
+import ru.serce.jnrfuse.struct.*;
+
+import java.nio.file.Path;
+import java.util.concurrent.CountDownLatch;
+
+public class MountSuccessFuseDecorator implements FuseFS {
```
Comment (overheadhunter) [inline src/main/java/org/cryptomator/frontend/fuse/mount/MountSuccessFuseDecorator.java:12]: can be package-private, can't it?

**Review by overheadhunter — APPROVED** [review]
LGTM except for two minor improvements that aren't deal-breakers.

File: src/main/java/org/cryptomator/frontend/fuse/mount/AbstractMounter.java (hunk 1 above)
Comment (infeo) [reply src/main/java/org/cryptomator/frontend/fuse/mount/AbstractMounter.java]: Fixed in 72684a2

File: src/main/java/org/cryptomator/frontend/fuse/mount/MountSuccessFuseDecorator.java (hunk 2 above)
Comment (infeo) [reply src/main/java/org/cryptomator/frontend/fuse/mount/MountSuccessFuseDecorator.java]: Fixed in 72684a2

**Review by overheadhunter — APPROVED** [review]

**Pull request merged**